src/nanoemoji/bitmap_tables.py

```diff
@@ -0,0 +1,269 @@
+            ),
+            y_offset=int(
+                round(line_ascent - 0.5 * (line_height - config.bitmap_resolution))
+            ),
+            line_height=line_height,
+            line_ascent=int(line_ascent),
+        )
+
+        print("line_height", metrics.line_height)
+        print("line_ascent", metrics.line_ascent)
+        print("metrics.x_offset", metrics.x_offset)
+        print("metrics.y_offset", metrics.y_offset)
+
+        return metrics
+
+
+# https://github.com/googlefonts/noto-emoji/blob/9a5261d871451f9b5183c93483cbd68ed916b1e9/third_party/color_emoji/emoji_builder.py#L53
+def _ppem(config: FontConfig, advance: int) -> int:
+    return int(config.bitmap_resolution * config.upem / advance)
+
+
+def _advance(ttfont: ttLib.TTFont, color_glyphs: Sequence[ColorGlyph]) -> int:
+    # let's go ahead and fail miserably if advances are not all the same
+    # proportional bitmaps can wait for a second pass :)
+    advances = {
+        ttfont["hmtx"].metrics[ttfont.getGlyphName(c.glyph_id)][0] for c in color_glyphs
+    }
+    assert len(advances) == 1, "Proportional bitmaps not supported yet"
+    return next(iter(advances))
+
+
+def _image_bytes(color_glyph: ColorGlyph) -> bytes:
+    with open(color_glyph.filename, "rb") as f:
+        return f.read()
+
+
+def _cbdt_record_size(image_format: int, image_data: bytes) -> int:
+    assert image_format == _CBDT_SMALL_METRIC_PNGS, "Unrecognized format"
+    return _CBDT_SMALL_METRIC_PNG_HEADER_SIZE + len(image_data)
+
+
+def _cbdt_bitmapdata_offsets(
+    image_format: int, color_glyphs: Sequence[ColorGlyph]
+) -> List[Tuple[int, int]]:
+    # TODO is this right? - feels dumb. But ... compile crashes if locations are unknown.
+    offsets = []
+    offset = _CBDT_HEADER_SIZE
+    for color_glyph in color_glyphs:
+        offsets.append(offset)
+        offset += _cbdt_record_size(image_format, _image_bytes(color_glyph))
+    offsets.append(offset)  # capture end of stream
+    return list(zip(offsets, offsets[1:]))
+
+
+def _cbdt_bitmap_data(
+    config: FontConfig, metrics: BitmapMetrics, image_data: bytes
+) -> CbdtBitmapFormat17:
+
+    bitmap_data = CbdtBitmapFormat17(b"", None)
+    bitmap_data.metrics = SmallGlyphMetrics()
+    bitmap_data.metrics.height = config.bitmap_resolution
+    bitmap_data.metrics.width = config.bitmap_resolution
+    # center within advance
+    bitmap_data.metrics.BearingX = metrics.x_offset
+    bitmap_data.metrics.BearingY = metrics.y_offset
+    bitmap_data.metrics.Advance = config.bitmap_advance
+    bitmap_data.imageData = image_data
+    return bitmap_data
+
+
+def make_sbix_table(
+    config: FontConfig,
+    ttfont: ttLib.TTFont,
+    color_glyphs: Sequence[ColorGlyph],
+):
+
+    sbix = ttLib.newTable("sbix")
+    ttfont[sbix.tableTag] = sbix
+
+    sbix.ppem = _ppem(config, _advance(ttfont, color_glyphs))
+    # TODO what goes here
+    sbix.ppi = 96
+
+    strike = SbixStrike()
+    strike.ppem = sbix.ppem + 8  # TODO magic #
```

Comment:
I'm puzzled by the + 8. Also, why sbix.ppem would be different from SbixStrike.ppem?

BTW, I am getting weird rendering in Safari Mac, the bitmaps look too small and offset towards the top:

![image](https://user-images.githubusercontent.com/6939968/150522105-05e7a090-3954-4d90-8da9-f985f05dda92.png)

In Chrome Mac, instead I just see AdobeBlank fallback when rendering sbix... :(